README
NAME
    Lingua::ZH::CCDICT - An interface to the CCDICT Chinese dictionary

SYNOPSIS
      use Lingua::ZH::CCDICT;

      my $dict = Lingua::ZH::CCDICT->new( storage => 'InMemory' );

DESCRIPTION
    This module provides a Perl interface to the CCDICT dictionary created
    by Thomas Chin. This dictionary is indexed by Unicode character number
    (traditional character), and contains information about these
    characters.

    CCDICT is released under a Creative Commons Attribution License (version
    2.5).

    The dictionary contains the following information, though not all
    information is avaialable for every character.

    * Radical number
        The number of the radical. Always available.

    * Index
        The total number of strokes minus the number of strokes in the
        radical. Always available.

    * Total stroke count
        The total number of strokes in the character.

    * Cangjie
        The Cangjie Chinese input system code.

    * Four Corner
        The Four Corner Chinese input system code.

    In addition, the dictionary contains English definitions (often multiple
    definitions), and romanizations for the character in different languages
    and systems. The romanizations available include Pinjim for Hakka,
    Jyutping for Cantonese and (Hanyu) Pinyin for Mandarin.

DICTIONARY BUGS
    The CCDICT dictionary is distributed by Thomas Chin in a simple but
    non-standard, textual ASCII-only format. I've tried to work around
    errors or ambiguities in the dictionary data, although there are
    probably still oddities lurking. Please send bug reports to me so I can
    figure out whether the error is in my code or the dictionary itself.

STORAGE
    This module is capable of parsing the CCDICT format file, and can also
    store the data in other formats (just Berkeley DB fo rnow).

    Each storage system is implemented via a module in the
    "Lingua::ZH::CCDICT::Storage::*" class hierarchy. All of these modules
    are subclasses of "Lingua::ZH::CCDICT" class, and implement its methods
    for searching the dictionary.

    In addition some storage classes may offer additional methods.

  Storage Subclasses
    The following storage subclasses are available:

    * Lingua::ZH::CCDICT::Storage::InMemory
        This class stores the entire parsed dictionary in memory.

    * Lingua::ZH::CCDICT::Storage::BerkeleyDB
        This class can convert the CCDICT source file to a set of BerkeleyDB
        files.

USAGE
    This module allows you to look up information in the dictionary based on
    a number of keys. These include the Unicode character (as a character,
    not its number), stroke count, radical number, and any of the various
    romanization systems.

METHODS
    This class provides the following methods.

  Lingua::ZH::CCDICT->new(...)
    This method always takes at least one parameter, "storage". This
    indicates what storage subclass to use. The current options are
    "InMemory" and "BerkeleyDB".

    Any other parameters given will be passed to the appropriate subclass's
    "new()" method.

  $dict->parse_source_file($filename)
    If you don't specify a file, then it will use the data file distributed
    with this module. This is probably what you want, unless you have a
    local copy of the dictionary that you want to work with. Note that the
    dictionary format has changes a fair bit between versions, so this
    probably won't work with much older or newer versions of the CCDICT
    data.

    This method is what does the real work of creating a dictionary. Note
    that if you are not using the InMemory storage subclass, you only need
    to parse the source file once, and then you can reuse the stored data.

MATCH METHODS
    When doing a lookup based on the romanization of a character, the tone
    is indicated with a number at the end of the syllable, as opposed to
    using the Unicode character combining the latin letter with the
    diacritic.

    In addition, lookups based on a Pinyin romanization should use the
    u-with-umlaut character (character 252 in Unicode) rather than two "u"
    characters.

    The return value for any lookup will be an object in a
    "Lingua::ZH::CCDICT::ResultSet" subclass.

    Result sets always return matches in ascending Unicode character order.

  $ccdict->match_unicode(@chars)
    This method matches on one or more Unicode characters. Unicode
    characters should be given as Perl characters (i.e. "chr(0x7D20)"), not
    as a number.

    This dictionary index uses *traditional* Chinese characters. Simplified
    character lookups will not work (but you could use "Encode::HanConvert"
    to convert simple to traditional first).

  $ccdict->match_radical(@numbers)
    Given a set of numbers, this method returns those characters containing
    the specified radical(s).

  $ccdict->match_index(@numbers)
    Given a set of numbers, this method returns those characters containing
    the specified index(es).

  $ccdict->match_stroke_count(@numbers)
    Given a set of numbers, this method returns those characters containing
    the specified number(s) of strokes.

  $ccdict->match_cangjie(@codes)
    Given a set of Cangjie codes, this method returns the character(s) for
    those code(s).

  $ccdict->match_four_corner(@codes)
    Given a set of Four Corner codes, this method returns the character(s)
    for those code(s).

  $ccdict->match_pinjim(@romanizations)
  $ccdict->match_jyutping(@romanizations)
  $ccdict->match_pinyin(@romanizations)
  $ccdict->all_characters()
    Returns a result set containing all of the characters in the dictionary.

  $ccdict->entry_count()
    Returns the number of entries in the dictionary

ENVIRONMENT VARIABLES
    There are several environment variables you can set to change this
    module's behavior.

    * CCDICT_DEBUG_SOURCE
        Causes a warning when bad data is enountered in the ccdict
        dictionary source. This is primarily useful if you want to find bugs
        in the dictionary itself.

    * CCDICT_VERBOSE
        Tells the module to give you progress reports when parsing the
        source file. These are sent to STDERR.

AUTHOR
    David Rolsky <autarch@urth.org>

COPYRIGHT
    Copyright (c) 2002-2007 David Rolsky. All rights reserved. This program
    is free software; you can redistribute it and/or modify it under the
    same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

    CCDICT is copyright (c) 1995-2006 Thomas Chin.

SEE ALSO
    Lingua::ZH::CEDICT - for converting between Chinese and English.

    Encode::HanConvert - for converting between simplified and traditional
    characters in various character sets.

    http://www.chinalanguage.com/dictionaries/CCDICT/ - the home of the
    CCDICT dictionary.



Changes
0.04  2007-11-26

- The 0.03 release did not declare Sub::Name as a prereq.


0.03  2007-11-25

* Updated to CCDICT 5.1.1. The data has changed a fair amount, and
  this means a number of API changes. Notably, most of the different
  romanizations are gone, and there is only each for (Hanyu) Pinyin,
  Hakka Pinjim, and Cantonese Jyutping.

- Unicode representations for pinyin with one vowel at the beginning
  of the syllable were not being created at all.

- Fixed an off-by-one error which caused this module to skip the last
  item in the dictionary. Reported by Scott Lanning. RT #27717.

- Changed obsolete() to is_obsolete(). The old version is still
  available, but deprecated.


0.02  2003-02-26

- Fixed a bug in turning fifth tone (no tone) pinyin syllables into
  Unicode.  The syllable was being turned into '', which is definitely
  not right ;)

- Fixed COPYRIGHT sections of docs to specify that the module is
  licensed under the same terms as Perl itself, not just the Artistic
  License.


0.01  2002-08-23 (from Taiwan)

- First release.  It seems to work but no doubt has bugs.



lib/Lingua/ZH/CCDICT/ResultItem.pm
package Lingua::ZH::CCDICT::ResultItem;

use strict;
use warnings;

use Params::Validate qw( validate SCALAR UNDEF ARRAYREF );

foreach my $item ( qw( unicode radical index stroke_count cangjie four_corner ) )
{
    no strict 'refs';
    *{ __PACKAGE__ . "::$item"} =
        sub { return unless exists $_[0]->{$item};
              $_[0]->{$item} };
}

foreach my $item ( qw( jyutping pinyin pinjim english ) )
{
    no strict 'refs';
    *{ __PACKAGE__ . "::$item"} =
        sub { return unless exists $_[0]->{$item};
              wantarray ? @{ $_[0]->{$item} } : $_[0]->{$item}[0] };

}


sub new
{
    my $class = shift;
    my %p = validate( @_,
                      { unicode      => { type => SCALAR },
                        radical      => { type => SCALAR },
                        index        => { type => SCALAR },
                        stroke_count => { type => SCALAR, optional => 1 },
                        cangjie      => { type => SCALAR, optional => 1 },
                        four_corner  => { type => SCALAR, optional => 1 },
                        jyutping     => { type => ARRAYREF, optional => 1 },
                        pinyin       => { type => ARRAYREF, optional => 1 },
                        pinjim       => { type => ARRAYREF, optional => 1 },
                        english      => { type => ARRAYREF, optional => 1 },
                      },
                    );

    return bless \%p, $class;
}


1;

__END__

=head1 NAME

Lingua::ZH::CCDICT::ResultItem - A single result from a dictionary search

=head1 SYNOPSIS

  print $item->unicode();

  print $_->syllable(), "\n" for $item->pinyin();

=head1 DESCRIPTION

Each individual result returned by a C<Lingua::ZH::CCDICT::ResultSet>
returns an object of this class.

=head1 METHODS

This class provides the following methods:

=head2 $item->unicode()

=head2 $item->radical()

=head2 $item->index()

=head2 $item->stroke_count()

=head2 $item->cangjie()

=head2 $item->four_corner()

These methods always return a single item when the requested data is
available or a false value if this item is not available.

=head2 $item->pinjim()

=head2 $item->jyutping()

=head2 $item->pinyin()

=head2 $item->english()

These methods represent data for which there may be multiple values.
In a list context, all values are returned. In a scalar context, only
the first value is returned. When the requested data is not
available, a false value is returned.

Romanizations are returned as C<Lingua::ZH::CCDICT::Romanization>
objects.

=head1 AUTHOR

David Rolsky <autarch@urth.org>

=head1 COPYRIGHT

Copyright (c) 2002-2007 David Rolsky. All rights reserved. This
program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

The full text of the license can be found in the LICENSE file included
with this module.

=cut


lib/Lingua/ZH/CCDICT/Storage/InMemory.pm
package Lingua::ZH::CCDICT::Storage::InMemory;

use strict;
use warnings;

use base 'Lingua::ZH::CCDICT';

use Params::Validate qw( validate SCALAR );

use Lingua::ZH::CCDICT::ResultItem;
use Lingua::ZH::CCDICT::ResultSet::Array;


sub new
{
    my $class = shift;
    my %p = validate( @_,
                      { file => { type => SCALAR, optional => 1 },
                      },
                    );

    my $self = bless {}, $class;

    $self->parse_source_file( $p{file} );

    return $self;
}

sub _real_add_entry
{
    my $self    = shift;
    my $unicode = shift;
    my $entry   = shift;

    foreach my $key ( keys %$entry )
    {
        next if $key eq 'english' || $key eq 'unicode';

        foreach my $val ( eval { @{ $entry->{$key} } } ? @{ $entry->{$key} } : $entry->{$key} )
        {
            # intentionally stringify to take advantage of
            # stringification overloading for romanizations
            push @{ $self->{$key}{"$val"} }, $unicode;
        }
    }

    $self->{unicode}{$unicode} = $entry;
    $self->{unicode}{$unicode}{unicode} = $unicode;
}

sub all_characters
{
    my $self = shift;

    return
        Lingua::ZH::CCDICT::ResultSet::Array->new
            ( results =>
              [ map { Lingua::ZH::CCDICT::ResultSet->new( %{ $self->{unicode}{$_} } ) }
                sort keys %{ $self->{unicode} } ]
            );
}

sub match_unicode
{
    my $self = shift;

    my %seen;
    my @chars = map { $seen{$_}++ ? () : $_ } sort @_;

    my @results;
    foreach my $char (@chars)
    {
        if ( exists $self->{unicode}{$char} )
        {
            push @results,
                Lingua::ZH::CCDICT::ResultItem->new( %{ $self->{unicode}{$char} } );
        }
    }

    return Lingua::ZH::CCDICT::ResultSet::Array->new( array => \@results );
}

sub _match
{
    my $self = shift;
    my $type = shift;

    return
        $self->match_unicode( map { @{ $self->{$type}{$_} } }
                              grep { exists $self->{$type}{$_} }
                              @_
                            );
}

sub entry_count
{
    return scalar keys %{ $_[0]->{unicode} };
}


1;

__END__

=head1 NAME

Lingua::ZH::CCDICT::Storage::InMemory - Store the dictionary in memory

=head1 USAGE

  use Lingua::ZH::CCDICT;

  my $dict = Lingua::ZH::CCDICT->new( storage => 'InMemory',
                                      file    => '/path/to/source/ccdict.txt',
                                    );

=head1 DESCRIPTION

This module stores the CCDICT dictionary in memory.  It uses a rather
large amount of memory in doing so (234MB on my system), but it is
fast.

The object can be serialized to disk and reloaded, which is quicker
than parsing the dictionary source repeatedly.

For serious usage, the C<Lingua::ZH::CCDICT::Storage::BerkeleyDB>
class is strongly recommended, since you can save the results of
parsing the CCDICT data between invocations of your code.

=head1 METHODS

This subclass offers no extra methods.

=head1 AUTHOR

David Rolsky <autarch@urth.org>

=head1 COPYRIGHT

Copyright (c) 2002-2007 David Rolsky. All rights reserved. This
program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

The full text of the license can be found in the LICENSE file included
with this module.

=cut


lib/Lingua/ZH/CCDICT.pm
package Lingua::ZH::CCDICT;

use strict;
use warnings;

use 5.006001;

use vars qw ($VERSION);

$VERSION = '0.04';

use File::Spec;
use Params::Validate qw(:all);
use Sub::Name qw( subname );
use Lingua::ZH::CCDICT::Romanization;
use Lingua::ZH::CCDICT::Romanization::Pinyin;

my %storage;
BEGIN { %storage = map { lc $_ => $_ } ( qw( InMemory XML BerkeleyDB ) ) }


sub new
{
    my $class = shift;
    my %p = validate_with( params => \@_,
                           spec   =>
                           { storage =>
                             { callbacks =>
                               { 'is a valid storage type' =>
                                 sub { $storage{ lc $_[0] } } },
                             },
                           },
                           allow_extra => 1,
                         );

    my $storage_class = __PACKAGE__ . '::Storage::' . $storage{ lc delete $p{storage} };

    eval "use $storage_class";
    die $@ if $@;

    return $storage_class->new(%p);
}

my %Ignore = map { $_ => 1 } qw( fUTF8 fCNS11643 fGB fBig5 );

my %CCDICTToInternal =
    ( fTotalStrokes => 'stroke_count',
      fCangjie      => 'cangjie',
      fFourCorner   => 'four_corner',
    );

my %RomanizationToInternal =
    ( fHakka        => 'pinjim',
      fCantonese    => 'jyutping',
      fMandarin     => 'pinyin',
    );

my %RomanizationType = map { $_ => 1 } values %RomanizationToInternal;

sub parse_source_file
{
    my $self = shift;
    my $file = shift;
    my $status_fh = $ENV{CCDICT_VERBOSE} ? \*STDERR : undef;
    my $lines_per = 5000;

    unless ( defined $file )
    {
        my $pack_as_file = File::Spec->catfile( split /::/, __PACKAGE__ );
        $pack_as_file .= '.pm';

        (my $dir = $INC{$pack_as_file}) =~ s/\.pm$//;

        $file = File::Spec->catfile( $dir, 'Data.pm' );
    }

    open my $fh, '<', $file
        or die "Cannot read $file: $!";

    my $last_char;
    my %entry;
    while (<$fh>)
    {
        chomp;

        next unless substr( $_, 0, 1 ) eq 'U';

        my ( $unicode, $type, $data ) = split /\t/, $_, 3;

        s/^\s+|\s+$//g for $unicode, $type, $data;

        next if $Ignore{$type};

        my ( $codepoint, $homograph ) = $unicode =~ /U\+([\dABCDEF]+)\.(\d)/;

        die "Bad line (line #$.):\n$_\n\n" unless $codepoint;

        # not sure how to handle this, to be honest.
        next if $homograph;

        # generate a real Unicode character in Perl
        my $unicode_char = chr( hex($codepoint) );

        $last_char = $unicode_char unless defined $last_char;

        # this relies on the fact that data for each char is grouped
        # together on consecutive lines in the ccdict.txt file.
        if ( $unicode_char ne $last_char )
        {
            $self->_add_entry( $last_char, { %entry, unicode => $last_char } );

            %entry = ();

            $last_char = $unicode_char;
        }

        if ( exists $CCDICTToInternal{$type} )
        {
            my $internal = $CCDICTToInternal{$type};

            $entry{$internal} = $data;
        }
        elsif ( exists $RomanizationToInternal{$type} )
        {
            my $internal = $RomanizationToInternal{$type};

            my $class = $self->_romanization_class($internal);

            my $is_obsolete = 0;
            foreach my $syl ( split /[\s;]+/, $data )
            {
                # Deal with various mysterious bits and mistakes
                next if $syl eq '}';
                next if $syl eq '(also)';
                next if $syl eq '(old)';

                next if $syl =~ /^p?\d+$/;

                next if $syl =~ /^\(coll/;

                $syl =~ s/^{?([^}]+)}$/$1/;
                $syl =~ s/{[^}]+}?$//;

                if ( $syl eq 'obs' )
                {
                    $is_obsolete = 1;
                    next;
                }

                my $romanized = $class->new( syllable => $syl,
                                             obsolete => $is_obsolete,
                                           );

                next unless $romanized;

                push @{ $entry{$internal} }, $romanized;
            }
        }
        elsif ( $type eq 'fR/S' )
        {
            my ( $radical, $index ) = split /\./, $data;

            $entry{radical} = $radical;
            $entry{index} = $index;
        }
        elsif ( $type eq 'fEnglish' )
        {
            $entry{english} =
                [ grep { defined && length } split /\s*\[\d\d?\]\s*/, $data ];
        }
        else
        {
            die "Invalid line: $_\n";
        }

        if (  ! ( $. % $lines_per ) && $status_fh )
        {
            print $status_fh "$. lines processed\n";
        }
    }

    $self->_add_entry( $last_char, { %entry, unicode => $last_char } );
}

sub _add_entry
{
    my $self    = shift;
    my $unicode = shift;
    my $entry   = shift;

    return unless defined $entry->{radical};

    $self->_real_add_entry( $unicode, $entry );
}

sub _romanization_class
{
    return
        ( $_[1] eq 'pinyin' ?
          'Lingua::ZH::CCDICT::Romanization::Pinyin' :
          'Lingua::ZH::CCDICT::Romanization'
        );
}

sub _is_romanization_type
{
    return $RomanizationType{ $_[1] };
}

sub _InternalTypes
{
    return values %CCDICTToInternal, values %RomanizationToInternal, 'index', 'radical';
}

# Some of these may be overridden in subclasses, but they provide an
# easy default
foreach my $type ( __PACKAGE__->_InternalTypes() )
{
    my $sub =
        __PACKAGE__->_is_romanization_type($type)
        ? sub { my $s = shift;
                $s->_match( $type => map { lc } @_ ) }
        : sub { my $s = shift;
                $s->_match( $type => @_ ) };

    my $sub_name = "match_$type";
    no strict 'refs';

    *{$sub_name} = subname $sub_name => $sub;
}


1;

__END__


=head1 NAME

Lingua::ZH::CCDICT - An interface to the CCDICT Chinese dictionary

=head1 SYNOPSIS

  use Lingua::ZH::CCDICT;

  my $dict = Lingua::ZH::CCDICT->new( storage => 'InMemory' );

=head1 DESCRIPTION

This module provides a Perl interface to the CCDICT dictionary created
by Thomas Chin. This dictionary is indexed by Unicode character number
(traditional character), and contains information about these
characters.

CCDICT is released under a Creative Commons Attribution License
(version 2.5).

The dictionary contains the following information, though not all
information is avaialable for every character.

=over 4

=item * Radical number

The number of the radical. Always available.

=item * Index

The total number of strokes minus the number of strokes in the
radical. Always available.

=item * Total stroke count

The total number of strokes in the character.

=item * Cangjie

The Cangjie Chinese input system code.

=item * Four Corner

The Four Corner Chinese input system code.

=back

In addition, the dictionary contains English definitions (often
multiple definitions), and romanizations for the character in
different languages and systems. The romanizations available include
Pinjim for Hakka, Jyutping for Cantonese and (Hanyu) Pinyin for
Mandarin.

=head1 DICTIONARY BUGS

The CCDICT dictionary is distributed by Thomas Chin in a simple but
non-standard, textual ASCII-only format. I've tried to work around
errors or ambiguities in the dictionary data, although there are
probably still oddities lurking. Please send bug reports to me so I
can figure out whether the error is in my code or the dictionary
itself.

=head1 STORAGE

This module is capable of parsing the CCDICT format file, and can also
store the data in other formats (just Berkeley DB fo rnow).

Each storage system is implemented via a module in the
C<Lingua::ZH::CCDICT::Storage::*> class hierarchy. All of these
modules are subclasses of C<Lingua::ZH::CCDICT> class, and implement
its methods for searching the dictionary.

In addition some storage classes may offer additional methods.

=head2 Storage Subclasses

The following storage subclasses are available:

=over 4

=item * Lingua::ZH::CCDICT::Storage::InMemory

This class stores the entire parsed dictionary in memory.

=item * Lingua::ZH::CCDICT::Storage::BerkeleyDB

This class can convert the CCDICT source file to a set of BerkeleyDB
files.

=back

=head1 USAGE

This module allows you to look up information in the dictionary based
on a number of keys. These include the Unicode character (as a
character, not its number), stroke count, radical number, and any of
the various romanization systems.

=head1 METHODS

This class provides the following methods.

=head2 Lingua::ZH::CCDICT->new(...)

This method always takes at least one parameter, "storage". This
indicates what storage subclass to use. The current options are
"InMemory" and "BerkeleyDB".

Any other parameters given will be passed to the appropriate
subclass's C<new()> method.

=head2 $dict->parse_source_file($filename)

If you don't specify a file, then it will use the data file
distributed with this module. This is probably what you want, unless
you have a local copy of the dictionary that you want to work
with. Note that the dictionary format has changes a fair bit between
versions, so this probably won't work with much older or newer
versions of the CCDICT data.

This method is what does the real work of creating a dictionary. Note
that if you are not using the InMemory storage subclass, you only need
to parse the source file once, and then you can reuse the stored data.

=head1 MATCH METHODS

When doing a lookup based on the romanization of a character, the tone
is indicated with a number at the end of the syllable, as opposed to
using the Unicode character combining the latin letter with the
diacritic.

In addition, lookups based on a Pinyin romanization should use the
u-with-umlaut character (character 252 in Unicode) rather than two "u"
characters.

The return value for any lookup will be an object in a
C<Lingua::ZH::CCDICT::ResultSet> subclass.

Result sets always return matches in ascending Unicode character
order.

=head2 $ccdict->match_unicode(@chars)

This method matches on one or more Unicode characters. Unicode
characters should be given as Perl characters (i.e. C<chr(0x7D20)>),
not as a number.

This dictionary index uses I<traditional> Chinese characters.
Simplified character lookups will not work (but you could use
C<Encode::HanConvert> to convert simple to traditional first).

=head2 $ccdict->match_radical(@numbers)

Given a set of numbers, this method returns those characters
containing the specified radical(s).

=head2 $ccdict->match_index(@numbers)

Given a set of numbers, this method returns those characters
containing the specified index(es).

=head2 $ccdict->match_stroke_count(@numbers)

Given a set of numbers, this method returns those characters
containing the specified number(s) of strokes.

=head2 $ccdict->match_cangjie(@codes)

Given a set of Cangjie codes, this method returns the character(s) for
those code(s).

=head2 $ccdict->match_four_corner(@codes)

Given a set of Four Corner codes, this method returns the character(s)
for those code(s).

=head2 $ccdict->match_pinjim(@romanizations)

=head2 $ccdict->match_jyutping(@romanizations)

=head2 $ccdict->match_pinyin(@romanizations)

=head2 $ccdict->all_characters()

Returns a result set containing all of the characters in the
dictionary.

=head2 $ccdict->entry_count()

Returns the number of entries in the dictionary

=head1 ENVIRONMENT VARIABLES

There are several environment variables you can set to change this
module's behavior.

=over 4

=item * CCDICT_DEBUG_SOURCE

Causes a warning when bad data is enountered in the ccdict dictionary
source. This is primarily useful if you want to find bugs in the
dictionary itself.

=item * CCDICT_VERBOSE

Tells the module to give you progress reports when parsing the source
file. These are sent to STDERR.

=back

=head1 AUTHOR

David Rolsky <autarch@urth.org>

=head1 COPYRIGHT

Copyright (c) 2002-2007 David Rolsky. All rights reserved. This
program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

The full text of the license can be found in the LICENSE file included
with this module.

CCDICT is copyright (c) 1995-2006 Thomas Chin.

=head1 SEE ALSO

Lingua::ZH::CEDICT - for converting between Chinese and English.

Encode::HanConvert - for converting between simplified and traditional
characters in various character sets.

http://www.chinalanguage.com/dictionaries/CCDICT/ - the home of the CCDICT
dictionary.

=cut

